gpflow/models/training_interface.py

```diff
@@ -0,0 +1,62 @@
+import abc
+import collections
+from typing import Callable, Optional, Tuple, TypeVar
+
+
+import tensorflow as tf
+import numpy as np
+
+InputData = tf.Tensor
+OutputData = tf.Tensor
+RegressionData = Tuple[InputData, OutputData]
+Data = TypeVar("Data", RegressionData, InputData)
+
+
+class TrainingInterface:
+    @abc.abstractmethod
+    def training_loss(self, *args, **kwargs) -> tf.Tensor:
+        """
+        Specify a training loss for this model
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def training_loss_closure(self, *args, **kwargs):
+        """
+        Return a closure with a training loss for this module
+        """
+        raise NotImplementedError
```

Comment:
I've just discussed with eric on slack about APIs with `(*args, **kwargs)`. They present a problem in that subclasses must also accept any number of arguments and keyword arguments to satisfy the contract. In particular, they have to have the signature `(*args: Any, **kwargs: Any)`.

You can think of it like this: Imagine you're working in a function which has signature
```python3
def foo(obj: TrainingInterface) -> None:
```
how do you work with it? I should be able to pass in whatever I want, for example
```
obj.training_loss(4, np.random, values='which', make=not 'sense')
```
This would fail if the runtime object doesn't support those arguments

Reply:
do you mean the last sentence?  TBH am happy to get rid of the base `TrainingInterface`, to abide by liskov principle.

Reply:
Aaaand it feels like we're back to square one. I would very much like to go with the thing that actually works in practice, for users, rather than the thing that makes the type system happy but makes it more cumbersome for users...

Reply:
@st-- type systems aren't just there for theoretical aesthetics. I would guess that breaking the type system always results in at least one test that you can write that fails. It translates into code that's buggy in some way, and bugs are cumbersome for users

Reply:
Python isn't a strongly-typed language, so from a `*args, **kwargs` signature you can't infer that you should be *allowed* to pass in whatever you want.

Reply:
Trying to force static typing on Python can result in unpythonic and cumbersome code. I don't think that's the right approach for a research toolbox that's supposed to be flexible and make it easy for GP researchers to implement their ideas.

Reply:
> Python isn't a strongly-typed language, so from a `*args, **kwargs` signature you can't infer that you should be _allowed_ to pass in whatever you want.

it's not a matter of it being statically typed (python is strongly typed, but is optionally statically typed), it's a fundamental design principle of OOP (LSP), which applies equally for dynamic and static typing

> 

Trying to force static typing on Python can result in unpythonic and cumbersome code. I don't think that's the right approach for a research toolbox that's supposed to be flexible and make it easy for GP researchers to implement their ideas.

I don't agree it's unpythonic. Static types are after all part of the standard library. Cumbersome? Yes, if overused. But overused and cumbersome are both subjective. Regarding a research toolbox, I'm not really clear yet on what that means. I hope time will tell. Flexible? I don't anticipate this will make it any more or less flexible.

Reply:
I suspect the concern here is the conceptual overhead of the design. I agree it's a headache and not something we should keep in the code base. I guess the discussion is what to do until we reach that point. I lean on the side of dealing with it safely until we can converge on a final design. I appreciate that's not the universal opinion.